x/archivedb/key.go

```diff
@@ -4,85 +4,70 @@
-// archivedb is an append only database).
+// MaxUint64 minus the desired Height.
 //
-// Any other property are to not serialized but they are useful when parsing a
-// keyInternal struct from the database
-type keyInternal struct {
-	key       []byte
-	height    uint64
-	isDeleted bool
+// All keys are prefixed with a VarUint64, which is the length of the keys, the
+// idea is to make reads even simpler, making it effectively a O(1) operation.
+// Before this change it would have been O(1) unless a another key existed with
+// the same prefix as the requested. By appending the length of the inner key as
+// the first bytes, it is not longer possible to query by a prefix (the full key
+// is needed to do a lookup) but all reads are a O(1).
+func newDBKey(key []byte, height uint64) []byte {
+	keyLen := len(key)
+	rawKeyMaxSize := keyLen + binary.MaxVarintLen64 + longLen
+	rawKey := make([]byte, rawKeyMaxSize)
+	offset := binary.PutUvarint(rawKey, uint64(keyLen))
+	offset += copy(rawKey[offset:], key)
+	binary.BigEndian.PutUint64(rawKey[offset:], ^height)
+	offset += longLen
+
+	return rawKey[0:offset]
 }
 
-// Creates a new Key struct with a given key and its height
-func newInternalKey(key []byte, height uint64) *keyInternal {
-	return &keyInternal{
-		key:       key,
-		isDeleted: false,
-		height:    height,
+// Takes a slice of bytes and returns a the inner key and the height
+func parseDBKey(rawKey []byte) ([]byte, uint64, error) {
+	rawKeyLen := len(rawKey)
+	reader := bytes.NewReader(rawKey)
+	keyLen, err := binary.ReadUvarint(reader)
+	if err != nil {
+		return nil, 0, err
 	}
-}
 
-func (k *keyInternal) Bytes() []byte {
-	keyLen := len(k.key)
-	bytes := make([]byte, keyLen+internalKeySuffixLen+internalKeyPrefixLen)
-	copy(bytes[0:], internalKeyPrefix)
-	copy(bytes[internalKeyPrefixLen:], k.key)
-	binary.BigEndian.PutUint64(bytes[internalKeyPrefixLen+keyLen:], math.MaxUint64-k.height)
-	if k.isDeleted {
-		bytes[internalKeyPrefixLen+keyLen+longLen] = 1
-	} else {
-		bytes[internalKeyPrefixLen+keyLen+longLen] = 0
+	if minInternalKeyLen >= rawKeyLen {
+		return nil, 0, ErrInsufficientLength
 	}
-	return bytes
-}
 
-// Takes a slice of bytes and returns a Key struct
-func parseKey(rawKey []byte) (*keyInternal, error) {
-	var key keyInternal
-	if internalKeySuffixLen+internalKeyPrefixLen >= len(rawKey) {
-		return nil, ErrInsufficientLength
+	key := make([]byte, keyLen)
+	readBytes, err := reader.Read(key)
+	if err != nil {
+		return nil, 0, err
 	}
 
-	keyWithPrefixLen := len(rawKey) - internalKeySuffixLen
+	if uint64(readBytes) != keyLen {
+		return nil, 0, database.ErrNotFound
+	}
 
-	key.key = make([]byte, keyWithPrefixLen-internalKeyPrefixLen)
-	key.height = math.MaxUint64 - binary.BigEndian.Uint64(rawKey[keyWithPrefixLen:])
-	key.isDeleted = rawKey[keyWithPrefixLen+longLen] == 1
-	copy(key.key, rawKey[internalKeyPrefixLen:internalKeyPrefixLen+keyWithPrefixLen])
+	inversedHeight := binary.BigEndian.Uint64(rawKey[rawKeyLen-longLen:])
```

Comment:
Can we add a comment for this? Or maybe a README section? I don't think it'll be immediately obvious to readers what this is doing, or why